raiden/utils/migrations/v18_to_v19.py

```diff
@@ -0,0 +1,172 @@
+        """Given a block number returns the hex representation of the blockhash"""
+        if block_number in self.mapping:
+            return self.mapping[block_number]
+
+        block_hash = self.web3.eth.getBlock(block_number)['hash']
+        block_hash = block_hash.hex()
+        self.mapping[block_number] = block_hash
+        return block_hash
+
+
+def _add_blockhash_to_state_changes(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to ContractReceiveXXX and ActionInitChain state changes"""
+    state_changes = storage.get_all_state_changes()
+    updated_state_changes = []
+    for state_change in state_changes:
+        data = json.loads(state_change.data)
+        affected_state_change = (
+            'raiden.transfer.state_change.ContractReceive' in data['_type'] or
+            'raiden.transfer.state_change.ActionInitChain' in data['_type']
+        )
+        if affected_state_change:
+            assert 'block_hash' not in data, 'v18 state changes cant contain blockhash'
+            block_number = int(data['block_number'])
+            data['block_hash'] = cache.get(block_number)
+
+            updated_state_changes.append(StateChangeRecord(
+                state_change_identifier=state_change.state_change_identifier,
+                data=json.dumps(data),
+            ))
+
+    storage.update_state_changes(updated_state_changes)
+
+
+def _add_blockhash_to_events(storage: SQLiteStorage, cache: BlockHashCache) -> None:
+    """Adds blockhash to all ContractSendXXX events"""
+    events = storage.get_all_event_records()
+    updated_events = []
+    for event in events:
+        data = json.loads(event.data)
+        if 'raiden.transfer.events.ContractSend' in data['_type']:
+            assert 'triggered_by_block_hash' not in data, 'v18 events cant contain blockhash'
+            # Get the state_change that triggered the event and if it has
+            # a block number get its hash. If not fall back to latest.
+            matched_state_changes = storage.get_statechanges_by_identifier(
+                from_identifier=event.state_change_identifier,
+                to_identifier=event.state_change_identifier,
+            )
+            result_length = len(matched_state_changes)
+            msg = 'multiple state changes should not exist for the same identifier'
+            assert result_length == 0 or result_length == 1, msg
+            block_hash = None
+            if result_length == 1:
+                statechange_data = json.loads(matched_state_changes[0])
+                if 'block_hash' in statechange_data:
+                    block_hash = statechange_data['block_hash']
+                elif 'block_number' in statechange_data:
+                    block_number = int(statechange_data['block_number'])
+                    block_hash = cache.get(block_number)
+
+            # else fallback to just using the latest blockhash
+            if not block_hash:
```

Comment:
Looking at the above code: If `result_length` is zero... then we don't fail, we just fallback to the "latest". I think this is wrong. The migration should fail if we don't manage to find the state change that triggered the event.

Reply:
I am not sure about it. The migration failing is a really fatal error and I would rather avoid it. This is why I added this fallback.

On the other hand .. when can this happen? When can we fail to find the state change that triggered the event? That can only happen if our database is somehow corrupt or missing data. So we would already be fucked.

Hmm ... I think I agree with you then. I tried to not fail here, but perhaps it only makes sense to fail if this happens as it's a rather fatal error.

Reply:
Not finding the corresponding state change must never happen. The event table must have a foreign key constraint and the identifier of the state change table must have an unique constraint for the id.

Also, this fallback is really bad, it completely breaks the continuity of the events. Older events would point to newer blocks, and the newer block would be different because `latest` is not part of the cache.

Reply:
> The event table must have a foreign key constraint and the identifier of the state change table must have an unique constraint for the id

 I already removed the fallback. 

Yes this is exactly what I thought when putting more thinking into it. And yes not only there should be a foreign key constraint ... there already is one :+1: 

Source [here](https://github.com/raiden-network/raiden/blob/d05bb74d70fa196bf14bf334546014e7316fb8f2/raiden/storage/utils.py#L29)